### ISSUE TITLE
Paper #667: Making Symbolic Execution Promising by Learning Aggressive State-Pruning Strategy

### ISSUE THREAD

**Issue by kstolee** [researchart/fse20#81]
This issue is for discussion related to PR #45 

Authors are applying for the *reusable*, *functional*, and *available* badges.

Author contact: @sooyoungcha

**Comment by kstolee** [comment]
Authors: Sooyoung Cha and Hakjoo Oh

**Comment by art-reviewer** [comment]
[Paper](https://github.com/sooyoungcha/fse20/blob/add2af7ea4f6d81740bbe2313fa2f1cb8993eefc/submissions/reusable/Homi/FSE20.pdf)

**Comment by kstolee** [comment]
@AnonReviewer626 and @anon-review We would appreciate your review till Monday morning AoE.

**Comment by AnonReviewer626** [comment]
Here is my review.

## Available

- No DOI, so we cannot grant this badge yet.
- In STATUS there is a dead link to https://github.com/kupl/HOMI_public, not sure what to make of this...But a GitHub project is not archival-quality because it can be deleted by the owner.

## Functional

I think we can grant this badge if some conditions are met. See requests and concerns under the appropriate heading.

### Documented

- The license appears to be that of KLEE. It reads at the end *"The KLEE software contains code written by third parties.  Such software will
have its own individual LICENSE.TXT file in the directory in which it appears.
This file will describe the copyrights, license, and restrictions which apply
to that code."*

Please clarify the license of this artifact.

### Consistent

I have some concerns here. In Algorithm 1, the pruning is probabilistic. I don't see reference to how you controlled the randomization in your experiments (I searched the paper for *seed*). You instead say "We repeated all experiments five times and reported the average results", but don't include variance so I cannot tell whether this was a statistically sound choice. I also don't see this as a tunable within the artifact. 

I do not have access to the computational resources necessary to reproduce the authors' results. However, I was able to install and run their VM for long enough to confirm that "everything seemed to be working" on the small benchmark they suggested.

**Requested changes**:
1. In the artifact, please clarify how randomness can be controlled for fully reproducible results.
2. I recommend you also update the paper to address this aspect of your work. If it's in there and I couldn't find it, you might wish to rework the prose a bit anyway.

### Complete

Based on the ~/Homi directory in the project, I believe everything described in the paper is included in the artifact.

### Exercisable

1. Unclear whether the complete source code is included in the artifact, since there is no (working) link to the Homi project source code nor description. See also my requests under Reusable.

2. I ran the enscript example and I am not sure if it worked. In particular, I specified a time budget of `15 seconds` and the program ran for 100 seconds before I Ctrl-C'd it. Please fix this bug and/or clarify+enforce any minimum time budget values.

## Reusable

I recommend we grant this with some minor tweaks.

### Documentation

The instructions for "How to run Homi on arbitary C open-source programs" are exemplary, but I have some questions.

1. What does the command `$  find . -executable -type f | xargs -I '{}' extract-bc '{}'` do in the enscript example?

2. The step "Generate the json file" needs more description. Is the behavior for enscript hardcoded? What if I need to tweak the json file? Please clarify the specification for the json file.

### Reuse and repurposing is facilitated

3. Please document how you extended Klee to implement Homi. Is it a plugin or a major overhaul? Did you use standard Klee techniques? Do you have a visualization of the updated architecture?

## Misc.

- In Algorithm 1 of the paper, the symbols for the path condition and for the case condition are similar enough that I cannot easily tell them apart.  Likewise the program P and probability P look quite similar. Perhaps consider an alternative font or different notation?
- In Algorithm 1, with the extra page for the camera-ready I suggest you add notes indicating that the algorithm has conditions for non-if statements as well. It might also be helpful to clarify, either in the caption or inline in the algorithm, whether the probabilistic pruning is standard.
- In Algorithm 1, you might consider moving the pruning until after the state processing to clarify that it is an implementation consideration to address real-world memory limitations.
- The paper says "(1) How promising each state is?" -- Should be "(1) How promising is each state?"

----------

CC @sooyoungcha

**Comment by anon-review** [comment]
✅ Functional

Needs revision for Available and Reusable badges.

## Review

❓**Available**

Cannot grant this badge without a DOI link. The artefact is "available" at a Google Drive link. To be awarded the **Available** badge, it needs to be publicly accessible at an archival repository (e.g., Zenodo), and have a DOI assigned to it.

Also, to be considered "available" this artefact needs to be under an open source license. The provided license appears to be that of KLEE. Please include a license for this artefact.

`STATUS.md` says the artefact is available at a GitHub link that results in a 404 error. `INSTALL.md` provides the valid Google Drive link. Please resolve this inconsistency.

✅ **Functional**

☑️ Documented

- Was able to download the VM image and run it using VirtualBox. I had to give the VM fewer resources than recommended, since I only have an 8GB memory machine available.
- Inconsistency in steps to run for arbitrary C programs? `~/Homi/README.md` says the following:

    **3)** Lastly **run Homi** with the search heuristic (nurs:cpicnt) on **enscript-1.6.6** for **5 hours** as:

```bash
$ cd ~/Homi/script
$ python3.7 Homi.py pgm_config/1enscript.json 1800 homi nurs:cpicnt 1
```

Looking at the `Homi.py` source, 1800 is in units of seconds? This is 30 minutes, not five hours. Did the authors mean 18000?

☑️ Consistent + Complete

All listed benchmarks are provided (4 smaller ones  + big programs in CoreUtils):

```bash
homi@homi-VM:~/Homi/benchmarks$ ls
combine-0.4.0  coreutils-8.31  gawk-3.1.4  grep-2.6  trueprint-5.4
```

Source code for running comparisons appears to be present.

☑️ Exercisable

**Running Homi**

Was able to run the Homi script. It terminated after a couple of hours, and I was able to produce the two pdfs `trueprint_cov.pdf` and `trueprint_statenum.pdf`

**Arbitrary C programs**

I was able to perform steps 1) and 2) for the enscript example in `reusable/README.md`. Step 3) had a possible inconsistency, see above.

I ran `Homi.py` with the provided arguments, only changing 1800 to a time limit of 10 seconds. This does not seem to have been enforced, or it was enforced only after several other steps took place.

`Homi.py` terminated with the following output:

```bash
Lines executed:17.78% of 3807
#############################################
################Time Out!!!!!################
#############################################
```

❓**Reusable**

Agree with @AnonReviewer626 that the json config format should be specified. This is my only request before awarding this badge.

**Comment by sooyoungcha** [comment]
We greatly appreciate the reviewers’ valuable comments.

---------------------------------------@AnonReviewer626-------------------------------------------------
[Q1]. Provide a DOI link for this artifact. 
We added a DOI link (https://doi.org/10.5281/zenodo.3902978) for this artifact in STATUS.md. We also updated the Github link (https://github.com/kupl/Homi_public) to be accessible.

[Q2]. Please clarify the license of this artifact. 
We update the license of this artifact. See LICENSE.txt.

[Q3]. About controlling the randomness in our experiments.
As reviewer mentioned, we did not try to control the randomness for fully reproducible results in our experiments. This is because when testing real-world programs, the randomness is inevitable. Instead of, we repeated all experiments five times and reported the average results. Additionally, we plan to report the variance of each technique in our revised paper.  

[Q4]. Is the complete code in this artifact? 
Yes, we include the complete source code in this artifact. 

[Q5]. Could you fix the bug about the time budget? 
We fixed the bug that Homi would run longer than a given small time budget (e.g., 15s). Now, Homi is performed during a given time budget. However, it is recommended to run Homi at least 30 minutes for testing real-world benchmarks.

[Q6]. What does the command “find . -executable -type f | xargs -I '{}' extract-bc '{}'” do in the enscript example?
This command extracts the LLVM bitcode from a build project. We refer to the following klee documentation. (https://klee.github.io/tutorials/testing-coreutils/)

[Q7]. Could you clarify the specification for the json file?
Yes, we update README.md by adding the description about the json file, where it contains 6 information of each benchmark.

[Q8]. How did you extend KLEE to implement Homi?
We extended the standard KLEE in order to prune the states based on the given n features and n-dimensional weight vectors. We mainly implemented it in the files 'Homi/klee/lib/Core/Executor.cpp' and 'Homi/klee/lib/Core/Executor.h'. That is, our pruning technique is not available in the original KLEE. 

Our new implementation for KLEE is able to prune states through three steps; the first step is to transform each state into a n-dimensional boolean vector when given n features. Second, it scores each state with a given weight vector. Third, it prunes the states based on the scores and the pruning ratio. 

---------------------------------------@anon-review-------------------------------------------------------
[Q9]. Provide a DOI link and include a license for this artifact. 
We added a DOI link (https://doi.org/10.5281/zenodo.3902978) for this artifact in STATUS.md. We also updated the Github link (https://github.com/kupl/Homi_public) to be accessible.

[Q10]. About inconsistency in steps to run for an arbitrary C program. 
This is our mistake. We updated “5 hours” (18,000s) to “30 minutes” (1,800s).  

[Q11]. Could you clarify the specification for the json file?
Yes, we update README.md by adding the description about the json file, where it contains 6 information of each benchmark.

**Comment by kstolee** [comment]
Related to the above comment, PR #142 was just merged

**Comment by AnonReviewer626** [comment]
@sooyoungcha I see that the GitHub link is now publicly accessible. But that project does not contain the actual implementation used in the paper, just the description files included in the FSE'20 artifact repository (i.e. this repo).

So at the moment, the only place I can find your code is in the VirtualBox image.

Is it possible for you to publish your code repository as well? At the moment the VBox image does not seem to contain a Git repo (maybe I'm missing something?), so it's hard for me to understand what files you added to KLEE and what changes you made.

**Comment by anon-review** [comment]
@sooyoungcha  Thank you for your response. I have edited my earlier comment to grant the Available badge.

Regarding @AnonReviewer626's comments, I agree that an explicit listing of changes made to KLEE would be nice to have. But my review is not contingent on this addition.

**Comment by sooyoungcha** [comment]
@AnonReviewer626 
These are two files that represent the main differences between the standard KLEE and KLEE implemented in Homi. We attached two files: gitdiff_Executor_cpp and gitdiff_Executor_h
We hope the files will help the reviewer understand what changes we made. 

[gitdiff_Executor_h.txt](https://github.com/researchart/fse20/files/4815003/gitdiff_Executor_h.txt)
[gitdiff_Executor_cpp.txt](https://github.com/researchart/fse20/files/4815005/gitdiff_Executor_cpp.txt)

In addition, a few files (/lib/Expr/ExprPPrinter.cpp or /tools/klee/main.cpp) have been slightly modified.

**Comment by AnonReviewer626** [comment]
@sooyoungcha I understand you made several changes. I am trying to understand whether you plan to publish a version-controlled repository tracking those changes, or whether you will only be providing the virtual machine?

**Comment by sooyoungcha** [comment]
@AnonReviewer626
Originally, we were planning to provide only the virtual machine.
However, we will also consider releasing a version-controlled repository based on the reviewers' comment that the release will help other researchers.

**Comment by AnonReviewer626** [comment]
@sooyoungcha 

I will approve the **functional** badge without the repo, but would prefer to see the version-controlled repository for the **reusable** badge. I view this as part of the guidelines for "reusable": *"Reuse and repurposing is facilitated"*.

If @anon-review disagrees we can discuss.

**Comment by sooyoungcha** [comment]
@AnonReviewer626
We just updated the KLEE we extended to implement Homi to the git repository (https://github.com/kupl/HOMI_public/). Now, if the other researchers look the commit (https://github.com/kupl/HOMI_public/commit/956f5110705dad1676d543fb7d421dd8050f7c94), they will know what changes we made compared to the standard KLEE.

**Comment by AnonReviewer626** [comment]
@sooyoungcha 

Thanks for making the KLEE updates. Looks really helpful to future researchers!

Your README says "The following command will run our tool, Homi with the given search heuristic, to test one of our benchmarks, trueprint-5.4, for an hour once."

I ran this, and got the following result:
![image](https://user-images.githubusercontent.com/66540393/85421668-b35e6b80-b542-11ea-856d-9dcc39f96402.png)

Have you updated the VM? I am still using the original VM you provided, so maybe that's the trouble.

**Comment by sooyoungcha** [comment]
@AnonReviewer626
[Q1]. Have you updated the VM? Where can I find this change? 
Yes, we update the VM. We also attach the git change file.

![diff_homi](https://user-images.githubusercontent.com/31407515/85429887-56060100-b5ba-11ea-8de5-5d2a2d6a732d.PNG)

Before running the command (python3.7 Homi.py gm_config/1trueprint.json 3600 homi nurs:md2u 1), we recommend that the reviewer cleans in the experiment directory (i.e., rm -rf /Homi/experiments/*).

[Q2].  Do you intend to share the other content from the VM (Homi.py, benchmarks, experiments, and script folders) in this repository as well?
Note that we basically decided to release Homi with VM becase the other researchers required a lot of effort (e.g.,  klee build, each benchmark build, and install the dependency) to install homi from scratch. 

However, we have updated the script folder containing Homi.py, one of our core implementations, to the git repository (https://github.com/kupl/HOMI_public).

**Comment by AnonReviewer626** [comment]
@sooyoungcha 

> Yes, we update the VM.

I'm sorry, I must be looking in the wrong place for it then. Can you share the new link?

> Note that we basically decided to release Homi with VM becase the other researchers required a lot of effort (e.g., klee build, each benchmark build, and install the dependency) to install homi from scratch.

I appreciate this distinction, and I want to clarify why I am asking about this:
- I think a VM or Dockerfile really helps with *reviewing* the artifact (thank you! I am glad I did not have to build all of the benchmarks and install dependencies!).
- But for *extending* it, having the relevant version-controlled source code is helpful. For example, if I were to ask one of my students to reproduce and extend your results as a class assignment, I would want them to start from the source code, not your VM. If there are any other scripts/drivers/etc. in the VM, would it be possible for you to duplicate them into the version-controlled repo? You don't need to include things like the benchmark source code, those are fine to leave in the VM.

**Comment by sooyoungcha** [comment]
@AnonReviewer626
This is the new link (https://drive.google.com/file/d/1ukvyUtVLJ0ie9knvsJKTe0Ww4K6U92H_/view), where it is in the git repository (https://github.com/kupl/HOMI_public/).

Currently, the Homi directory in VM is the same as Homi_public (https://github.com/kupl/HOMI_public/) except benchmark directory and installed dependency.

**Comment by gousiosg** [comment]
@AnonReviewer626 would you reconsider the `reusable` badge after all the updates?

**Comment by AnonReviewer626** [comment]
@sooyoungcha Thank you for all of the changes. Looks good.

I am happy to recommend **available, functional, and reusable** badges.

@gousiosg Since @anon-review also recommended those badges, I think we are all set.

**Comment by kstolee** [comment]
Thanks everyone! 

@sooyoungcha please be sure to cite your artifact using the DOI in the references for your paper. Congratulations!

**Comment by sooyoungcha** [comment]
@AnonReviewer626 and @anon-review

Thank you for all the valuable comments!
